sklearn/neighbors/_base.py

```diff
@@ -1096,15 +1094,15 @@ def fit(self, X, y):
 
         Parameters
         ----------
-        X : {array-like, sparse matrix, BallTree, KDTree}
+        X : {array-like, sparse matrix, BallTree, KDTree, NeighborsBase}
```

Comment:
I don't understand this. When are we fitting this estimator with X=estimator?

Reply:
This also looks a bit strange to me. I realized that we accept fitting an estimator with `X` being an estimator due to these lines:

https://github.com/scikit-learn/scikit-learn/blob/db673eb72ec2e7cbe5c30f50f97719a3bdf2f1b1/sklearn/neighbors/_base.py#L377-L382

In fact, this test:

https://github.com/scikit-learn/scikit-learn/blob/db673eb72ec2e7cbe5c30f50f97719a3bdf2f1b1/sklearn/neighbors/tests/test_neighbors.py#L98-L114

is fitting the `nbrs` estimator setting `X` to `nbrs_fid`, which is another estimator.

Reply:
This behavior is not consistent with the rest of scikit-learn's API, and it is not well documented, as briefly discussed in https://github.com/scikit-learn/scikit-learn/issues/10463#issue-288121351. I think the main goal was to avoid computing multiple times the nearest neighbors, when fitting multiple nearest-neighbors-based estimators.

Reusing precomputed nearest neighbors graphs was implemented in #10482 with an API more in line with the rest of scikit-learn. So, ideally, I would be in favor of removing the ability to use BallTree/KDTree/NeighborsBase in X, but this would need annoying deprecation cycles. We can also keep it as it is, but I would not advertise it much in the documentation.

Reply:
> This behavior is not consistent with the rest of scikit-learn's API, and it is not well documented, as briefly discussed in [#10463 (comment)](https://github.com/scikit-learn/scikit-learn/issues/10463#issue-288121351). I think the main goal was to avoid computing multiple times the nearest neighbors, when fitting multiple nearest-neighbors-based estimators.
> 
> Reusing precomputed nearest neighbors graphs was implemented in #10482 with an API more in line with the rest of scikit-learn. So, ideally, I would be in favor of removing the ability to use BallTree/KDTree/NeighborsBase in X, but this would need annoying deprecation cycles. We can also keep it as it is, but I would not advertise it much in the documentation.

I agree with you, I would be happy to remove the ability to use `sklearn.neighbors.BallTree`, `sklearn.neighbors.KDTree` and `sklearn.neighbors.NeighborsBase` in `X` to be consistent with the API. Nevertheless, this would lead to several deprecation cycles that should be avoided.

Thus, I have removed them from the API documentation since using `sklearn.neighbors.BallTree`, `sklearn.neighbors.KDTree` and `sklearn.neighbors.NeighborsBase` in `X` is not a typical use case.